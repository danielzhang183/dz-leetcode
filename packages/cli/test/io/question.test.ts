import { describe, expect, it } from 'vitest'
import { loadQuestions } from '../fixtures'

const questions = await loadQuestions()

export type RawQuestion = string | number

export interface UnresolvedUsage {
  category: string
  tagMap: Record<string, RawQuestion[]>
  questions: RawQuestion[]
}

describe('questions', () => {
  it('load questions', async () => {
    expect(questions).toMatchSnapshot('loadedQuestions')
  })

  it('no', async () => {
    const categories: Record<string, Record<string, (string | number)[]>> = {}

    for (const question of questions!) {
      const {
        category = 'unknown-category',
        tag = 'unknown-tag',
        questions,
      } = question
      if (!categories[category])
        categories[category] = {}
      if (!categories[category][tag])
        categories[category][tag] = []

      categories[category][tag].push(...questions)
    }

    expect(categories).toMatchInlineSnapshot(`
      {
        "algorithms": {
          "binary-search": [
            35,
            69,
            153,
          ],
          "sort": [
            15,
            912,
          ],
        },
        "structures": {
          "array": [
            1,
            26,
            27,
          ],
          "linked-list": [
            19,
            21,
            141,
          ],
        },
      }
    `)

    const usages: UnresolvedUsage[] = Object.entries(categories)
      // only check deps with more then 1 version in use
      .filter(i => Object.keys(i[1]).length > 1)
      // sort by the number of versions
      .sort((a, b) => Object.keys(b[1]).length - Object.keys(a[1]).length)
      .map(([category, tagMap]) => ({ category, tagMap, questions: Object.values(tagMap).flat(1) }))

    expect(usages).toMatchInlineSnapshot(`
      [
        {
          "category": "structures",
          "questions": [
            1,
            26,
            27,
            19,
            21,
            141,
          ],
          "tagMap": {
            "array": [
              1,
              26,
              27,
            ],
            "linked-list": [
              19,
              21,
              141,
            ],
          },
        },
        {
          "category": "algorithms",
          "questions": [
            15,
            912,
            35,
            69,
            153,
          ],
          "tagMap": {
            "binary-search": [
              35,
              69,
              153,
            ],
            "sort": [
              15,
              912,
            ],
          },
        },
      ]
    `)
  })
})
