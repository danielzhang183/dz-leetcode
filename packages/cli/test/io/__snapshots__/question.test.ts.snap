// Vitest Snapshot v1

exports[`questions > load questions > loadedQuestions 1`] = `
[
  {
    "category": "structures",
    "questions": [
      1,
      26,
      27,
    ],
    "tag": "array",
  },
  {
    "category": "structures",
    "questions": [
      19,
      21,
      141,
    ],
    "tag": "linked-list",
  },
  {
    "category": "algorithms",
    "questions": [
      15,
      912,
    ],
    "tag": "sort",
  },
  {
    "category": "algorithms",
    "questions": [
      35,
      69,
      153,
    ],
    "tag": "binary-search",
  },
]
`;
